quinn/src/delay_queue.rs

```diff
@@ -0,0 +1,442 @@
+                .iter()
+                .all(|level| level.slots.iter().all(|x| x.is_none())),
+            "lower levels are empty"
+        );
+        debug_assert!(
+            self.levels[level].slots[..slot].iter().all(Option::is_none),
+            "lower slots in this level are empty"
+        );
+
+        // Advance into the slot
+        self.next_tick = slot_start(self.next_tick, level, slot);
+
+        // Drain it into the lower levels, if possible
+        if level > 0 {
+            while let Some(timer) = self.levels[level].slots[slot].take() {
+                let next = self.timers[timer.0].next;
+                self.levels[level].slots[slot] = next;
+                if let Some(next) = next {
+                    self.timers[next.0].prev = None;
+                }
+                self.list_unlink(timer);
+                self.schedule(timer);
+            }
+        }
+    }
+
+    /// Link `timer` from the slot associated with its expiry
+    fn schedule(&mut self, timer: Timer) {
+        debug_assert_eq!(
+            self.timers[timer.0].next, None,
+            "timer isn't already scheduled"
+        );
+        debug_assert_eq!(
+            self.timers[timer.0].prev, None,
+            "timer isn't already scheduled"
+        );
+        let (level, slot) = timer_index(self.next_tick, self.timers[timer.0].expiry);
+        // Insert `timer` at the head of the list in the target slot
+        let head = self.levels[level].slots[slot];
+        self.timers[timer.0].next = head;
+        if let Some(head) = head {
+            self.timers[head.0].prev = Some(timer);
+        }
+        self.levels[level].slots[slot] = Some(timer);
+    }
+
+    /// Lower bound on when the next timer will expire, if any
+    pub fn next_timeout(&self) -> Option<u64> {
+        for level in 0..LEVELS {
+            let start = ((self.next_tick >> (level * LOG_2_SLOTS)) & (SLOTS - 1) as u64) as usize;
+            for slot in start..SLOTS {
+                if self.levels[level].slots[slot].is_some() {
+                    return Some(slot_start(self.next_tick, level, slot));
+                }
+            }
+        }
+        None
+    }
+
+    /// Register a timer that will yield `value` at `timeout`
+    pub fn insert(&mut self, timeout: u64, value: T) -> Timer {
+        let timer = Timer(self.timers.insert(TimerState {
+            expiry: timeout.max(self.next_tick),
+            prev: None,
+            next: None,
+            value,
+        }));
+        self.schedule(timer);
+        timer
+    }
+
+    /// Adjust `timer` to expire at `timeout`
+    pub fn reset(&mut self, timer: Timer, timeout: u64) {
+        self.unlink(timer);
+        self.timers[timer.0].expiry = timeout.max(self.next_tick);
+        self.schedule(timer);
+    }
+
+    /// Cancel `timer`
+    #[cfg(test)]
+    pub fn remove(&mut self, timer: Timer) -> T {
+        self.unlink(timer);
+        self.timers.remove(timer.0).value
```

Comment:
Does `Slab::remove()` ever shrink the allocation if the amount of required items gets reduced? Without that a lot of timer memory might hang around if one had a large amount of connections, and then moved to a lower amount. A "real" intrusive list would avoid that. Or one could do a manual shrink if e.g. the utilized capacity gets below a certain percentage..

Reply:
Good thought. Added logic to shrink if the number of timers is less than 1/16th the capacity.